hazelcast-client-legacy/src/main/java/com/hazelcast/client/spi/impl/listener/ClientNonSmartListenerService.java

```diff
@@ -23,100 +23,134 @@
+                registrations.remove(key);
+            } catch (Exception e) {
+                throw new HazelcastException("Listener with id " + userRegistrationId + " could not be removed", e);
+            }
+            return true;
         }
-        return true;
     }
 
     @Override
-    public void memberAdded(MembershipEvent membershipEvent) {
-        Address ownerConnectionAddress = client.getClientClusterService().getOwnerConnectionAddress();
-        if (membershipEvent.getMember().getAddress().equals(ownerConnectionAddress)) {
-            executionService.executeInternal(new Runnable() {
-                @Override
-                public void run() {
+    public void connectionAdded(Connection connection) {
+        executionService.executeInternal(new Runnable() {
+            @Override
+            public void run() {
+                synchronized (listenerRegLock) {
                     for (ClientRegistrationKey registrationKey : registrations.keySet()) {
-                        invoke(registrationKey);
+                        try {
+                            ClientEventRegistration registration = invoke(registrationKey);
+                            registrations.put(registrationKey, registration);
+                        } catch (Exception e) {
+                            logger.warning("Listener " + registrationKey + " could not be added ");
+                        }
                     }
                 }
-            });
-        }
+            }
+        });
     }
 
     @Override
-    public void memberRemoved(MembershipEvent membershipEvent) {
-        Address ownerConnectionAddress = client.getClientClusterService().getOwnerConnectionAddress();
-        if (membershipEvent.getMember().getAddress().equals(ownerConnectionAddress)) {
-            for (ClientEventRegistration registration : registrations.values()) {
-                removeEventHandler(registration.getCallId());
+    public void connectionRemoved(Connection connection) {
+        synchronized (listenerRegLock) {
+            for (Map.Entry<ClientRegistrationKey, ClientEventRegistration> entry : registrations.entrySet()) {
+                removeEventHandler(entry.getValue().getCallId());
             }
         }
     }
 
-    @Override
-    public void memberAttributeChanged(MemberAttributeEvent memberAttributeEvent) {
-        //ignore
+    //For Testing
+    public Collection<ClientEventRegistration> getActiveRegistrations(String uuid) {
+        synchronized (listenerRegLock) {
+            ClientEventRegistration registration = registrations.get(new ClientRegistrationKey(uuid));
+            if (registration == null) {
+                return Collections.EMPTY_LIST;
+            }
+            LinkedList<ClientEventRegistration> activeRegistrations = new LinkedList<ClientEventRegistration>();
+            if (getEventHandler(registration.getCallId()) != null) {
+                activeRegistrations.add(registration);
```

Comment:
Multiple calls to getActiveRegistrations with the same uuid will increase activeRegistrations on each call. Is this what 's desired from this method?